pkg/kubelet/nodeconfig/controller.go

```diff
@@ -0,0 +1,270 @@
+	cc.setConfigOK(fmt.Sprintf("using current (UID: %q)", curUID), "passed all checks", apiv1.ConditionTrue)
+
+	return cur
+}
+
+// initSync launches the sync loop that watches the API server for new configuration
+func (cc *NodeConfigController) initSync() {
+	if cc.client != nil {
+		infof("starting sync loop")
+		go func() {
+			fieldselector := fmt.Sprintf("metadata.name=%s", cc.nodename)
+
+			// Add some randomness to resync period, which can help avoid controllers falling into lock-step
+			minResyncPeriod := 30 * time.Second
+			factor := rand.Float64() + 1
+			resyncPeriod := time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+
+			_, ctl := cache.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options metav1.ListOptions) (kuberuntime.Object, error) {
+						return cc.client.Core().Nodes().List(metav1.ListOptions{
+							FieldSelector: fieldselector,
+						})
+					},
+					WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+						return cc.client.Core().Nodes().Watch(metav1.ListOptions{
+							FieldSelector:   fieldselector,
+							ResourceVersion: options.ResourceVersion,
+						})
+					},
+				},
+				&apiv1.Node{},
+				resyncPeriod,
+				cache.ResourceEventHandlerFuncs{
+					AddFunc: func(n interface{}) {
+						cc.onAddNodeEvent(n)
+					},
+					UpdateFunc: func(old interface{}, nw interface{}) {
+						cc.onWatchNodeEvent(nw)
+					},
+					DeleteFunc: func(n interface{}) {
+						return // do nothing, the node is probably being torn down
+					},
+				})
+
+			// since the Kubelet exits once it detects new config, no need to expose the stop channel
+			stop := make(chan struct{})
+			ctl.Run(stop)
+			return
+		}()
+	}
+}
+
+// onAddEvent syncs any status that was set before the Node was registered,
+// then calls onWatchNodeEvent to complete event handling.
+func (cc *NodeConfigController) onAddNodeEvent(n interface{}) {
+	cc.syncConfigOK()
+	cc.onWatchNodeEvent(n)
+}
+
+// onWatchNodeEvent checks for new config and downloads it if necessary.
+// If filesystem issues prevent proper operation of syncNodeConfig, a fatal error occurs.
+func (cc *NodeConfigController) onWatchNodeEvent(n interface{}) {
+	node, ok := n.(*apiv1.Node)
+	if !ok {
+		errorf("failed to cast watched object to Node, couldn't handle event")
+		return
+	}
+
+	// check the Node and download any new config
+	if updated, cause, err := cc.syncNodeConfig(node); err != nil {
+		errorf("failed to sync node config, error: %v", err)
+		// Update the ConfigOK status to reflect that we failed to sync, and so we don't know which configuration
+		// the user actually wants us to use. In this case, we just continue using the currently-in-use configuration.
+		cc.setConfigOK(cc.configOK.Message, fmt.Sprintf("failing to sync, desired config unclear, cause: %s", cause), apiv1.ConditionUnknown)
+		cc.syncConfigOK()
+	} else if updated {
+		infof("config updated, Kubelet will restart to begin using new config")
+		os.Exit(0)
```

Comment:
Yes, it's immediate. We might want to be more graceful. I think part of bringing this to beta will be figuring out what is and isn't compatible with this way of restarting. That said, the old implementation of dynamic configuration did the same thing, and we didn't see problems with the e2e node tests that rely on dynamic config.